File: shared/src/commonMain/kotlin/org/mlm/miniter/project/Keyframe.kt
package org.mlm.miniter.project

import androidx.compose.animation.core.*
import androidx.compose.ui.graphics.Color
import kotlinx.serialization.Serializable

@Serializable
data class Keyframe(
    val param: String,
    val offsetMs: Long,
    val value: Float,
    val easing: Easing = Easing.Linear
)

@Serializable
enum class Easing { Linear, EaseIn, EaseOut, EaseInOut }

@Serializable
data class KeyframeCurve(
    val keyframes: List<Keyframe> = emptyList()
) {
    fun evaluate(param: String, timeMs: Long): Float? {
        val relevant = keyframes.filter { it.param == param }.sortedBy { it.offsetMs }
        if (relevant.isEmpty()) return null
        if (relevant.size == 1) return relevant[0].value

        val t = timeMs
        if (t <= relevant.first().offsetMs) return relevant.first().value
        if (t >= relevant.last().offsetMs) return relevant.last().value

        for (i in 0 until relevant.lastIndex) {
            val a = relevant[i]
            val b = relevant[i + 1]
            if (t in a.offsetMs..b.offsetMs) {
                val range = (b.offsetMs - a.offsetMs).toFloat()
                if (range <= 0f) return b.value
                val progress = ((t - a.offsetMs).toFloat() / range).coerceIn(0f, 1f)
                val eased = applyEasing(a.easing, progress)
                return a.value + (b.value - a.value) * eased
            }
        }
        return relevant.last().value
    }

    fun toKeyframesSpec(param: String, durationMs: Long): KeyframesSpec<Float>? {
        val relevant = keyframes.filter { it.param == param }.sortedBy { it.offsetMs }
        if (relevant.size < 2) return null

        return keyframes {
            relevant.forEach { kf ->
                kf.value at kf.offsetMs.toInt() using when (kf.easing) {
                    Easing.Linear -> LinearEasing
                    Easing.EaseIn -> CubicBezierEasing(0.42f, 0f, 1f, 1f)
                    Easing.EaseOut -> CubicBezierEasing(0f, 0f, 0.58f, 1f)
                    Easing.EaseInOut -> CubicBezierEasing(0.42f, 0f, 0.58f, 1f)
                }
            }
            durationMillis = durationMs.toInt()
        }
    }
}

private fun applyEasing(easing: Easing, t: Float): Float = when (easing) {
    Easing.Linear -> t
    Easing.EaseIn -> t * t * t
    Easing.EaseOut -> 1f - (1f - t) * (1f - t) * (1f - t)
    Easing.EaseInOut -> if (t < 0.5f) 4f * t * t * t else 1f - (-2f * t + 2f) * (-2f * t + 2f) * (-2f * t + 2f) / 2f
}

object KeyframeParams {
    const val OPACITY = "opacity"
    const val VOLUME = "volume"
    const val TRANSFORM_SCALE = "transform.scale"
    const val TRANSFORM_TRANSLATE_X = "transform.translate_x"
    const val TRANSFORM_TRANSLATE_Y = "transform.translate_y"
    const val TRANSFORM_ROTATE = "transform.rotate"
    const val TEXT_POSITION_X = "text.position_x"
    const val TEXT_POSITION_Y = "text.position_y"
    const val TEXT_FONT_SIZE = "text.font_size"

    const val MASK_FEATHER = "mask.feather"
    const val MASK_SCALE = "mask.scale"
    const val MASK_TRANSLATE_X = "mask.translate_x"
    const val MASK_TRANSLATE_Y = "mask.translate_y"
    const val MASK_ROTATE = "mask.rotate"
    const val MASK_SHAPE_LEFT = "mask.shape.left"
    const val MASK_SHAPE_TOP = "mask.shape.top"
    const val MASK_SHAPE_RIGHT = "mask.shape.right"
    const val MASK_SHAPE_BOTTOM = "mask.shape.bottom"
    const val MASK_SHAPE_CENTER_X = "mask.shape.center_x"
    const val MASK_SHAPE_CENTER_Y = "mask.shape.center_y"
    const val MASK_SHAPE_RADIUS_X = "mask.shape.radius_x"
    const val MASK_SHAPE_RADIUS_Y = "mask.shape.radius_y"

    fun filterParam(index: Int, name: String) = "filter.$index.$name"
    fun maskParam(index: Int, name: String) = "mask.$index.$name"
}

data class ParamDef(
    val key: String,
    val displayName: String,
    val range: ClosedFloatingPointRange<Float>,
    val default: Float,
    val steps: Int = 0,
    val format: (Float) -> String,
    val color: Color = Color.Gray,
)

private fun fmtPct(v: Float) = "${(v * 100).toInt()}%"
private fun fmt1d(v: Float) = "${(v * 10).toInt() / 10f}x"
private fun fmtDeg(v: Float) = "${v.toInt()}°"
private fun fmt2d(v: Float) = "${(v * 100).toInt() / 100f}"

val MASK_PARAM_DEFS: Map<String, ParamDef> = listOf(
    ParamDef("feather", "Feather", 0f..0.2f, default = 0f, format = ::fmtPct, color = Color(0xFFE040FB)),
    ParamDef("scale", "Scale", 0.1f..5f, default = 1f, 24, format = ::fmt1d, color = Color(0xFFE040FB)),
    ParamDef("translate_x", "Pan X", -1f..1f, default = 0f, format = ::fmtPct, color = Color(0xFFE040FB)),
    ParamDef("translate_y", "Pan Y", -1f..1f, default = 0f, format = ::fmtPct, color = Color(0xFFE040FB)),
    ParamDef("rotate", "Rotate", -180f..180f, default = 0f, format = ::fmtDeg, color = Color(0xFFE040FB)),
    ParamDef("left", "Left", 0f..1f, default = 0.1f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("top", "Top", 0f..1f, default = 0.1f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("right", "Right", 0f..1f, default = 0.9f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("bottom", "Bottom", 0f..1f, default = 0.9f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("center_x", "Center X", 0f..1f, default = 0.5f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("center_y", "Center Y", 0f..1f, default = 0.5f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("radius_x", "Radius X", 0f..1f, default = 0.3f, format = ::fmtPct, color = Color(0xFFCE93D8)),
    ParamDef("radius_y", "Radius Y", 0f..1f, default = 0.3f, format = ::fmtPct, color = Color(0xFFCE93D8)),
).associateBy { it.key }

val ALL_PARAMS: List<ParamDef> = listOf(
    ParamDef(KeyframeParams.OPACITY, "Opacity", 0f..1f, default = 1f, format = ::fmtPct, color = Color(0xFF5B8DEF)),
    ParamDef(KeyframeParams.VOLUME, "Volume", 0f..2f, default = 1f, 19, format = ::fmtPct, color = Color(0xFF4CAF50)),
    ParamDef(KeyframeParams.TRANSFORM_SCALE, "Scale", 0.5f..3f, default = 1f, 24, format = ::fmt1d, color = Color(0xFFFF9800)),
    ParamDef(KeyframeParams.TRANSFORM_TRANSLATE_X, "Pan X", -1f..1f, default = 0.0f, format = ::fmtPct, color = Color(0xFFFF9800)),
    ParamDef(KeyframeParams.TRANSFORM_TRANSLATE_Y, "Pan Y", -1f..1f, default = 0.0f, format = ::fmtPct, color = Color(0xFFFF9800)),
    ParamDef(KeyframeParams.TRANSFORM_ROTATE, "Rotate", -180f..180f, default = 0f, format = ::fmtDeg, color = Color(0xFFFF9800)),
    ParamDef(KeyframeParams.TEXT_POSITION_X, "Text X", 0f..1f, default = 0.5f, format = ::fmtPct, color = Color(0xFF26A69A)),
    ParamDef(KeyframeParams.TEXT_POSITION_Y, "Text Y", 0f..1f, default = 0.9f, format = ::fmtPct, color = Color(0xFF26A69A)),
    ParamDef(KeyframeParams.TEXT_FONT_SIZE, "Font Size", 12f..120f, default = 24f, 26, format = { "${it.toInt()}sp" }, color = Color(0xFF26A69A)),
    ParamDef(KeyframeParams.MASK_FEATHER, "Mask Feather", 0f..100f, default = 0f, format = { "${it.toInt()}px" }, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SCALE, "Mask Scale", 0.1f..5f, default = 1f, format = ::fmt1d, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_TRANSLATE_X, "Mask X", -1f..1f, default = 0f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_TRANSLATE_Y, "Mask Y", -1f..1f, default = 0f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_ROTATE, "Mask Rotate", -180f..180f, default = 0f, format = ::fmtDeg, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_LEFT, "Shape Left", 0f..1f, default = 0.25f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_TOP, "Shape Top", 0f..1f, default = 0.25f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_RIGHT, "Shape Right", 0f..1f, default = 0.75f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_BOTTOM, "Shape Bottom", 0f..1f, default = 0.75f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_CENTER_X, "Center X", 0f..1f, default = 0.5f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_CENTER_Y, "Center Y", 0f..1f, default = 0.5f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_RADIUS_X, "Radius X", 0f..0.5f, default = 0.25f, format = ::fmtPct, color = Color(0xFF7E57C2)),
    ParamDef(KeyframeParams.MASK_SHAPE_RADIUS_Y, "Radius Y", 0f..0.5f, default = 0.25f, format = ::fmtPct, color = Color(0xFF7E57C2)),
)

val ALL_PARAMS_BY_KEY: Map<String, ParamDef> = ALL_PARAMS.associateBy { it.key }

fun defaultOf(key: String): Float = ALL_PARAMS_BY_KEY[key]?.default ?: 0f

fun paramDefOrUnknown(key: String): ParamDef = ALL_PARAMS_BY_KEY[key] ?: ParamDef(
    key = key,
    displayName = key,
    range = 0f..1f,
    default = 0f,
    format = ::fmt2d,
)

File: rust/crates/miniter-domain/src/keyframe.rs
use crate::time::MediaDuration;
use serde::{Deserialize, Serialize};

pub mod param {
    pub const OPACITY: &str = "opacity";
    pub const VOLUME: &str = "volume";
    pub const TRANSFORM_SCALE: &str = "transform.scale";
    pub const TRANSFORM_TRANSLATE_X: &str = "transform.translate_x";
    pub const TRANSFORM_TRANSLATE_Y: &str = "transform.translate_y";
    pub const TRANSFORM_ROTATE: &str = "transform.rotate";
    pub const TEXT_POSITION_X: &str = "text.position_x";
    pub const TEXT_POSITION_Y: &str = "text.position_y";
    pub const TEXT_FONT_SIZE: &str = "text.font_size";

    pub const MASK_FEATHER: &str = "mask.feather";
    pub const MASK_SCALE: &str = "mask.scale";
    pub const MASK_TRANSLATE_X: &str = "mask.translate_x";
    pub const MASK_TRANSLATE_Y: &str = "mask.translate_y";
    pub const MASK_ROTATE: &str = "mask.rotate";
    pub const MASK_SHAPE_LEFT: &str = "mask.shape.left";
    pub const MASK_SHAPE_TOP: &str = "mask.shape.top";
    pub const MASK_SHAPE_RIGHT: &str = "mask.shape.right";
    pub const MASK_SHAPE_BOTTOM: &str = "mask.shape.bottom";
    pub const MASK_SHAPE_CENTER_X: &str = "mask.shape.center_x";
    pub const MASK_SHAPE_CENTER_Y: &str = "mask.shape.center_y";
    pub const MASK_SHAPE_RADIUS_X: &str = "mask.shape.radius_x";
    pub const MASK_SHAPE_RADIUS_Y: &str = "mask.shape.radius_y";
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
#[derive(Default)]
pub enum Easing {
    #[default]
    Linear,
    EaseIn,
    EaseOut,
    EaseInOut,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Keyframe {
    pub param: String,
    pub offset: MediaDuration,
    pub value: f32,
    #[serde(default)]
    pub easing: Easing,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct KeyframeCurve {
    pub keyframes: Vec<Keyframe>,
}

impl KeyframeCurve {
    pub fn evaluate(&self, param: &str, local_time: MediaDuration) -> Option<f32> {
        let relevant: Vec<&Keyframe> = self.keyframes.iter().filter(|k| k.param == param).collect();

        if relevant.is_empty() {
            return None;
        }

        if relevant.len() == 1 {
            return Some(relevant[0].value);
        }

        let t = local_time.as_micros();
        if t <= relevant[0].offset.as_micros() {
            return Some(relevant[0].value);
        }
        if t >= relevant[relevant.len() - 1].offset.as_micros() {
            return Some(relevant[relevant.len() - 1].value);
        }

        for i in 0..relevant.len() - 1 {
            let a = relevant[i];
            let b = relevant[i + 1];
            let t_a = a.offset.as_micros();
            let t_b = b.offset.as_micros();

            if t >= t_a && t <= t_b {
                let range = (t_b - t_a) as f32;
                if range <= 0.0 {
                    return Some(b.value);
                }
                let progress = ((t - t_a) as f32 / range).clamp(0.0, 1.0);
                let eased = apply_easing(a.easing, progress);
                return Some(a.value + (b.value - a.value) * eased);
            }
        }

        Some(relevant[relevant.len() - 1].value)
    }

    pub fn insert_sorted(&mut self, kf: Keyframe) -> usize {
        if let Some(pos) = self
            .keyframes
            .iter()
            .position(|k| k.param == kf.param && k.offset == kf.offset)
        {
            let (param, offset) = (kf.param.clone(), kf.offset);
            self.keyframes[pos] = kf;
            self.sort_by_offset();
            return self
                .keyframes
                .iter()
                .position(|k| k.param == param && k.offset == offset)
                .unwrap_or(0);
        }
        let idx = self
            .keyframes
            .iter()
            .position(|k| k.offset > kf.offset)
            .unwrap_or(self.keyframes.len());
        self.keyframes.insert(idx, kf);
        idx
    }

    /// Sort curves by offset. Deserialized projects (JSON/Kotlin) may carry
    /// unsorted keyframes, which breaks `evaluate`'s interval scan.
    pub fn sort_by_offset(&mut self) {
        self.keyframes.sort_by_key(|k| k.offset);
    }
}

pub fn ease_linear(t: f32) -> f32 {
    t
}

pub fn ease_in(t: f32) -> f32 {
    t * t * t
}

pub fn ease_out(t: f32) -> f32 {
    1.0 - (1.0 - t).powi(3)
}

pub fn ease_in_out(t: f32) -> f32 {
    if t < 0.5 {
        4.0 * t * t * t
    } else {
        1.0 - (-2.0 * t + 2.0).powi(3) / 2.0
    }
}

fn apply_easing(easing: Easing, t: f32) -> f32 {
    match easing {
        Easing::Linear => t,
        Easing::EaseIn => ease_in(t),
        Easing::EaseOut => ease_out(t),
        Easing::EaseInOut => ease_in_out(t),
    }
}

#[cfg(test)]
mod keyframe_curve_tests {
    use super::*;

    #[test]
    fn insert_sorted_dedupes_same_param_offset() {
        let mut curve = KeyframeCurve::default();
        for v in [0.5, 0.8] {
            curve.insert_sorted(Keyframe {
                param: "volume".into(),
                offset: MediaDuration::from_micros(1_000_000),
                value: v,
                easing: Easing::Linear,
            });
        }
        assert_eq!(curve.keyframes.len(), 1);
        assert_eq!(curve.keyframes[0].value, 0.8);
    }

    #[test]
    fn out_of_order_inserts_stay_evaluable() {
        let mut curve = KeyframeCurve::default();
        for (off, val) in [(3_000_000, 3.0), (1_000_000, 1.0), (2_000_000, 2.0)] {
            curve.insert_sorted(Keyframe {
                param: "volume".into(),
                offset: MediaDuration::from_micros(off),
                value: val,
                easing: Easing::Linear,
            });
        }
        let offsets: Vec<i64> = curve.keyframes.iter().map(|k| k.offset.as_micros()).collect();
        assert_eq!(offsets, vec![1_000_000, 2_000_000, 3_000_000]);
        assert_eq!(
            curve.evaluate("volume", MediaDuration::from_micros(1_500_000)),
            Some(1.5)
        );
    }
}
